Ergonomics improvements

ejmahler opened the pull request:
I've been using RustFFT to make my own Discrete Cosine Transform crate: https://github.com/ejmahler/rust_dct

It' still WIP but it's coming along nicely. Based on my experience I think a few more improvements can be made to rustfft's API

- Make FFTnum depend on Debug. This is critical for debugging the workings of generic code that uses rustFFT. when developing rustfft it's simple enough to add it when you need it, and remove it before committing, but it's not that simple when rustfft is an external crate. This shouldn't interfere with any future fixed precision plans, since those should implement debug.
- Rename Planner to FFTplanner -- this is just a naming clarity thing
- Depend on the `num_complex` and `num_traits` crates instead of `num`. This greatly reduces the amount of stuff to download/compile when compiling rustfft. Adding dependencies isn't a breaking change afaik, so if we decide we need other num crates, we can add them individually later
- Re-export the num_traits and num_complex crates. This is so that other crates can `use rustfft::num_complex::Complex;` and be confident they're using the same version of Complex as rustfft itself, without having to manually match up version numbers

Based on my experience so far, these changes will make rustfft easier to use

Commit:
Depend on num-traits and num-complex instead of num

Commit:
FFTnum now requires Debug

Commit:
renamed Planner to FFTPlanner

awelkie commented:
All of these changes seem good except for adding the requirement of `Debug` to `FFTnum`. I don't see why it's necessary. If you're hacking on some generic code that uses RustFFT, then couldn't you just add the constraint on that code while debugging? I don't see why it needs to be part of the `FFTnum` trait.

Commit:
Added "is_inverse()" trait

ejmahler commented:
I've added one more commit.

You can compute all discrete cosine transform types by converting the problem to a FFT and then doing pre-processing or post-processing. But they're generally only clearly defined for forward FFTs. I'm sure inverse FFTs could be used too if you worked out the formula.

But at the very minimum, you need to know whether the inner FFT is computing inverse or forward FFTs, otherwise it will silently output garbage data. Right now there's no way to tell, so I added a "IsInverse" trait, with a single method, "is_inverse()" which returns true if this instance returns inverse FFTs, and false otherwise.

While I was implementing this, I realized that this actually fixes a bug in MixedRadix, GoodThomas, and MixedRadixDoublebutterfly: Before, if the user passed in one inverse and one forward FFT to the constructor, they would silently get incorrect results. After this change, we can now assert that the two inner FFTs are both inverse or both forward.

ejmahler commented:
To answer your question about FFTnum + Debug:

The goal of this change is to improve quality-of-life for developers who use rustFFT from generic code.

In particular, this: 

>  If you're hacking on some generic code that uses RustFFT, then couldn't you just add the constraint on that code while debugging?

is easier said than done. The most common way a user of rust FFT would write generic code is to write something like:

`fn do_thing<T: rustfft::FFTnum>(&mut[T]) -> Complex<T>`

If they write several different functions like this, THEN they realize that they need debugging, they have to add a debug constraint to each of those functions, a painful process. The other option is for them to write their own trait that depends on FFTnum, and to add Debug to that. This is what my DCT library is currently doing -- it defines a "DCTnum" trait that just requires FFTnum and Debug.

But I think this second approach raises the barrier to entry for understanding other libraries, because now users have to keep more traits in their head than just one. But FFTnum's goal in my mind is to make it so the user only needs to keep one trait in their head, and requiring library writers to define their own traits defeats this goal.

I decided to check out what other popular rust computation libraries do -- 

- `cgmath` tends to define Debug for all their custom types, but they tend not to require debug anywhere on generic parameters to their functions/structs -- see https://github.com/brendanzab/cgmath/blob/master/src/quaternion.rs
- The `ndarray` crate's NdFloat trait seems to have a very similar goal to our FFTnum trait, and they depend on Debug: https://github.com/bluss/rust-ndarray/blob/master/src/linalg_traits.rs
- The `rulinalg` crate defines debug on all its structs, but not on its traits
- The `dft` crate depends on num::Float, which dodesn't depend on Debug.

It looks like some do, but most don't, depend on Debug. BUT, the only library I could find that actually defines a floating number trait like FFTnum does depend on Debug.

I think it will make RustFFT easier to use. But I can absolutely see arguments for wanting RustFFT to be simpler, and have fewer dependencies. So I'll leave it up to you. If you decide you don't want it in, I'll revert the commit.

awelkie commented on src/algorithm/butterflies.rs:
```diff
@@ -696,23 +760,10 @@ mod unit_tests {
-	use num::Zero;
-
-	#[test]
-	fn test_butterfly2() {
```
Why did you take out this test?

ejmahler replied:
Before, it was different than all the other butterfly instances because its constructor didn't take the `inverse` parameter. So it had to be separate. But, because of the addition of the IsInverse trait, Butterfly2 needs to know whether it was created as an inverse or not, meaning it now takes inverse in the constructor.

So now, it's identical to all the others, and can be tested via the macro. If you check line 783 of the additions, you can see it tested there.

awelkie commented:
Yeah, I don't think `Debug` belongs in a public trait if it doesn't have to be there. The rest of the changes look good, though.

Commit:
Revert "FFTnum now requires Debug"

This reverts commit e0ba38387e0fadd8088f40148d003ff4cebac3b1.

ejmahler commented:
Reverted
